pkg/storage/chunk/client/congestion/controller.go

```diff
@@ -0,0 +1,219 @@
+package congestion
+
+import (
+	"context"
+	"errors"
+	"io"
+	"math"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/grafana/loki/pkg/storage/chunk/client"
+)
+
+// AIMDController implements the Additive-Increase/Multiplicative-Decrease algorithm which is used in TCP congestion avoidance.
+// https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease
+type AIMDController struct {
+	inner client.ObjectClient
+
+	retrier Retrier
+	hedger  Hedger
+	metrics *Metrics
+
+	limiter       *rate.Limiter
+	backoffFactor float64
+	upperBound    rate.Limit
+}
+
+func NewAIMDController(cfg Config) *AIMDController {
+	lowerBound := rate.Limit(cfg.Controller.AIMD.Start)
+	upperBound := rate.Limit(cfg.Controller.AIMD.UpperBound)
+
+	if lowerBound <= 0 {
+		lowerBound = 1
+	}
+
+	if upperBound <= 0 {
+		// set to infinity if not defined
+		upperBound = rate.Limit(math.Inf(1))
+	}
+
+	backoffFactor := cfg.Controller.AIMD.BackoffFactor
+	if backoffFactor <= 0 {
+		// AIMD algorithm calls for halving rate
+		backoffFactor = 0.5
+	}
+
+	return &AIMDController{
+		limiter:       rate.NewLimiter(lowerBound, int(lowerBound)),
+		backoffFactor: backoffFactor,
+		upperBound:    upperBound,
+	}
+}
+
+func (a *AIMDController) Wrap(client client.ObjectClient) client.ObjectClient {
+	a.inner = client
+	return a
+}
+
+func (a *AIMDController) withRetrier(r Retrier) Controller {
+	a.retrier = r
+	return a
+}
+
+func (a *AIMDController) withHedger(h Hedger) Controller {
+	a.hedger = h
+	return a
+}
+
+func (a *AIMDController) withMetrics(m *Metrics) Controller {
+	a.metrics = m
+
+	a.updateLimitMetric()
+	return a
+}
+
+func (a *AIMDController) PutObject(ctx context.Context, objectKey string, object io.ReadSeeker) error {
+	return a.inner.PutObject(ctx, objectKey, object)
+}
+
+func (a *AIMDController) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, int64, error) {
+	// Only GetObject implements congestion avoidance; the other methods are either non-idempotent which means they
+	// cannot be retried, or are too low volume to care about
+
+	// TODO(dannyk): use hedging client to handle requests, do NOT hedge retries
+
+	rc, sz, err := a.retrier.Do(
+		func(attempt int) (io.ReadCloser, int64, error) {
+			a.metrics.requests.Add(1)
+
+			// in retry
+			if attempt > 0 {
+				a.metrics.retries.Add(1)
+			}
+
+			// apply back-pressure while rate-limit has been exceeded
+			for {
```

Comment:
What do you think about
```suggestion
			for !a.limiter.Allow() {
```
instead? It saves the `continue` and `break`.